src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs

```diff
@@ -0,0 +1,3144 @@
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo(j As IEnumerable(Of Integer))
+            Dim i As IEnumerable(Of Integer) = Enumerable.Range(1, 10)
+            
+            For Each c In [|i|]
+            Next
+            [|i|].ToHashSet()
+            
+            For Each c2 In [|j|]
+            Next
+            [|j|].ToHashSet()
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVB.VerifyAnalyzerAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestInvocationWithForEachLoop2()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> j)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        foreach (var c in [|i|].Select(m => m + 1).Where(m => m != 100))
+        {
+        }
+
+        [|i|].ToLookup(x => x);
+
+        foreach (var c2 in [|j|].Select(m => m + 1).Where(m => m != 100))
+        {
+        }
+
+        [|j|].ToImmutableArray();
+
+        var k = Enumerable.Range(1, 10).ToArray();
+        foreach (var c3 in k.Select(m => m + 1).Where(m => m != 100))
+        {
+        }
+        k.ToImmutableArray();
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(csharpCode);
+
+            var vbCode = @"
+Imports System.Collections.Generic
+Imports System.Collections.Immutable
+Imports System.Linq
+
+Namespace Ns
+    Public Class Hoo
+        Public Sub Goo(j As IEnumerable(Of Integer))
+            Dim i As IEnumerable(Of Integer) = Enumerable.Range(1, 10)
+            
+            For Each c In [|i|].Select(Function(m) m + 1).Where(Function(m) m <> 100)
+            Next
+            [|i|].ToLookUp(Function(x) x)
+            
+            For Each c2 In [|j|].Select(Function(m) m + 1).Where(Function(m) m <> 100)
+            Next
+            [|j|].ToHashSet()
+
+            Dim k = Enumerable.Range(1, 10).ToArray
+
+            For Each c3 In k.Select(Function(m) m + 1).Where(Function(m) m <> 100)
+            Next
+            k.ToImmutableArray()
+        End Sub
+    End Class
+End Namespace
+";
+            await VerifyVB.VerifyAnalyzerAsync(vbCode);
+        }
+
+        [Fact]
+        public async Task TestInvocationWithForEachLoop3()
+        {
+            var csharpCode = @"
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+public class Bar
+{
+    public void Sub(IEnumerable<int> j)
+    {
+        IEnumerable<int> i = Enumerable.Range(1, 10);
+        foreach (var c in [|i|].Select(m => m + 1).Where(m => m != 100))
+        {
+            [|i|].Where(x => x != 100).ToImmutableList();
+        }
+
+        [|i|].Select(k => k + 1).Skip(100).First();
+
+        foreach (var c2 in [|j|].Select(m => m + 1).Where(m => m != 100))
+        {
+            [|j|].Where(x => x != 100).ToImmutableSortedDictionary(m => m, n => n);
```

Comment:
NIT: i think if we could squiggle under the method which is causing the multiple enumeration it would be very useful for the user for fixing the violation

Reply:
My thought is if we point to the local/parameter reference location, user would directly try-cast/covert the underlying variable to fix it.

If reported under the whole invocation,
```
void Bar(IEnumerable a, IEnumerable b)
{
[|a.Count()|];
[|a.SequencialEqual(b)|];
}
```
Then this is be clear for `SequencialEqual`

Reply:
> If reported under the whole invocation,
> ```
> void Bar(IEnumerable a, IEnumerable b)
> {
> [|a.Count()|];
> [|a.SequencialEqual(b)|];
> }
> ```
> Then this is be clear for `SequencialEqual`

Yes that looks more clear